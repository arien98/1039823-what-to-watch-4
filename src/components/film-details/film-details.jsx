import React from "react";
import PropTypes from "prop-types";
import Tabs from "../tabs/tabs.jsx";
import {TabsType} from "../../common.js";
import MoreFilmsAlike from "../more-films-alike/more-films-alike.jsx";
import {connect} from "react-redux";
import {ActionCreator} from "../../reducer/screen/screen.js";
import {getCurrentTab, getShowedFilmId} from "../../reducer/screen/selectors.js";
import {getFilms} from "../../reducer/data/selectors.js";

const FilmDetails = (props) => {
  const {filmId, filmsData, onFilmCardClick, onTabClick, currentTab} = props;

  const filmData = filmsData.find((element) => {
    return element.id.toString() === filmId.toString();
  });

  const {
    title,
    genre,
    releaseDate,
    poster,
    bigPoster,
  } = filmData;

  return (
    <>
      <section className="movie-card movie-card--full">
        <div className="movie-card__hero">
          <div className="movie-card__bg">
            <img src={bigPoster} alt={title} />
          </div>

          <h1 className="visually-hidden">WTW</h1>

          <header className="page-header movie-card__head">
            <div className="logo">
              <a href="main.html" className="logo__link">
                <span className="logo__letter logo__letter--1">W</span>
                <span className="logo__letter logo__letter--2">T</span>
                <span className="logo__letter logo__letter--3">W</span>
              </a>
            </div>

            <div className="user-block">
              <div className="user-block__avatar">
                <img src="img/avatar.jpg" alt="User avatar" width="63" height="63" />
              </div>
            </div>
          </header>

          <div className="movie-card__wrap">
            <div className="movie-card__desc">
              <h2 className="movie-card__title">{title}</h2>
              <p className="movie-card__meta">
                <span className="movie-card__genre">{genre}</span>
                <span className="movie-card__year">{releaseDate}</span>
              </p>

              <div className="movie-card__buttons">
                <button className="btn btn--play movie-card__button" type="button">
                  <svg viewBox="0 0 19 19" width="19" height="19">
                    <use xlinkHref="#play-s"></use>
                  </svg>
                  <span>Play</span>
                </button>
                <button className="btn btn--list movie-card__button" type="button">
                  <svg viewBox="0 0 19 20" width="19" height="20">
                    <use xlinkHref="#add"></use>
                  </svg>
                  <span>My list</span>
                </button>
                <a href="add-review.html" className="btn movie-card__button">Add review</a>
              </div>
            </div>
          </div>
        </div>

        <div className="movie-card__wrap movie-card__translate-top">
          <div className="movie-card__info">
            <div className="movie-card__poster movie-card__poster--big">
              <img src={poster} alt={title} width="218" height="327" />
            </div>

            <div className="movie-card__desc">
              <nav
                className="movie-nav movie-card__nav"
              >
                <ul className="movie-nav__list">

                  {Object.values(TabsType).map((element, i) => {
                    return (
                      <li
                        key={Math.random + i}
                        className={`movie-nav__item ${currentTab === element ? `movie-nav__item--active` : ``}`}
                      >
                        <a
                          href="#"
                          className="movie-nav__link"
                          data-tab={element}
                          onClick={onTabClick}
                        >
                          {element}
                        </a>
                      </li>
                    );
                  })}

                </ul>
              </nav>

              <Tabs
                filmData={filmData}
                currentTab={currentTab}
              />

            </div>
          </div>
        </div>
      </section>

      <MoreFilmsAlike
        filmsAlike={filmsData
          .filter((it) => it.genre === filmData.genre)
          .filter((it) => it.id !== filmData.id)
          .slice(0, 4)}
        onFilmCardClick={onFilmCardClick}
      />
    </>
  );
};

FilmDetails.propTypes = {
  filmId: PropTypes.number.isRequired,
  filmsData: PropTypes.arrayOf(PropTypes.object).isRequired,
  onFilmCardClick: PropTypes.func.isRequired,
  onTabClick: PropTypes.func.isRequired,
  currentTab: PropTypes.string.isRequired,
};

const mapStateToProps = (state) => ({
  currentTab: getCurrentTab(state),
  filmId: getShowedFilmId(state),
  filmsData: getFilms(state),
});

const mapDispatchToProps = (dispatch) => ({
  onTabClick(evt) {
    evt.preventDefault();
    const tab = evt.target.dataset.tab;
    dispatch(ActionCreator.changeTab(tab));
  },
});

export {FilmDetails};

export default connect(mapStateToProps, mapDispatchToProps)(FilmDetails);
