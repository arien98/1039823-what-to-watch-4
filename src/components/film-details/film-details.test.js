import React from "react";
import renderer from "react-test-renderer";
import {FilmDetails} from "../film-details/film-details.jsx";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";

const mockStore = configureStore([]);

describe(`film deatails`, () => {
  test(`it renders correctly`, () => {

    const props = {
      onFilmCardClick: () => {},
      onTabClick: () => {},
      currentTab: `Overview`,
      filmData: {
        bigPoster: `img/johnny-english.jpg`,
        description: `A hardened CIA operative finds himself at the mercy of a precocious 9-year-old girl, having been sent undercover to surveil her family.`,
        director: `Anthony Mann`,
        genre: `criminal`,
        id: 47209,
        poster: `img/no-country-for-old-men.jpg`,
        ratingCount: 448,
        ratingLevel: `Very good`,
        ratingScore: 4.4,
        releaseDate: 1947,
        runtime: `0h 25m`,
        starring: [`Heinz Herald`, `Richard Weil`, `Anthony Mann`, `Erich von Stroheim`, `Charlie Hunnam`],
        title: `Dardjeeling Limited`,
      },
      filmsData: [
        {
          id: 2,
          title: `Bohemian Rhapsody`,
          poster: `img/bohemian-rhapsody.jpg`,
        },
        {
          id: 3,
          title: `Macbeth`,
          poster: `img/macbeth.jpg`,
        },
        {
          id: 4,
          title: `Aviator`,
          poster: `img/aviator.jpg`,
        },
      ],
    };

    const store = mockStore({});

    const tree = renderer
      .create(
          <Provider store={store}>
            <FilmDetails {...props} />
          </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

