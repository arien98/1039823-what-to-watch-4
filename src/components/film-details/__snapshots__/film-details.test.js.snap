// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`film deatails it renders correctly 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
    style={
      Object {
        "backgroundColor": "white",
      }
    }
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="Dardjeeling Limited"
          src="img/johnny-english.jpg"
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <header
        className="page-header movie-card__head"
      >
        <div
          className="logo"
        >
          <a
            className="logo__link"
            href="/"
            onClick={[Function]}
          >
            <span
              className="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              className="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              className="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <div
          className="user-block"
        >
          <a
            className="user-block__link"
            href="/login"
            onClick={[Function]}
          >
            Sign in
          </a>
        </div>
      </header>
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Dardjeeling Limited
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              criminal
            </span>
            <span
              className="movie-card__year"
            >
              1947
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="14"
                viewBox="0 0 18 14"
                width="18"
              >
                <use
                  xlinkHref="#in-list"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            <a
              className="btn movie-card__button"
              href="/films/1/review"
              onClick={[Function]}
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="Dardjeeling Limited"
            height="327"
            src="img/no-country-for-old-men.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <nav
            className="movie-nav movie-card__nav"
          >
            <ul
              className="movie-nav__list"
            >
              <li
                className="movie-nav__item movie-nav__item--active"
              >
                <a
                  className="movie-nav__link"
                  data-tab="Overview"
                  href="#"
                  onClick={[Function]}
                >
                  Overview
                </a>
              </li>
              <li
                className="movie-nav__item "
              >
                <a
                  className="movie-nav__link"
                  data-tab="Details"
                  href="#"
                  onClick={[Function]}
                >
                  Details
                </a>
              </li>
              <li
                className="movie-nav__item "
              >
                <a
                  className="movie-nav__link"
                  data-tab="Reviews"
                  href="#"
                  onClick={[Function]}
                >
                  Reviews
                </a>
              </li>
            </ul>
          </nav>
          <div
            className="movie-rating"
          >
            <div
              className="movie-rating__score"
            >
              4.4
            </div>
            <p
              className="movie-rating__meta"
            >
              <span
                className="movie-rating__level"
              >
                Normal
              </span>
              <span
                className="movie-rating__count"
              >
                448
                 rating
              </span>
            </p>
          </div>
          <div
            className="movie-card__text"
          >
            <p>
              A hardened CIA operative ercy of a precocious 9-year-old girl, having been sent undercover to surveil her family.
            </p>
            <p
              className="movie-card__director"
            >
              <strong>
                Director: 
                Anthony Mann
              </strong>
            </p>
            <p
              className="movie-card__starring"
            >
              <strong>
                Starring: 
                Heinz Herald, Anthony Mann, Erich von Stroheim, Charlie Hunnam
                and other
              </strong>
            </p>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <div
        className="catalog__movies-list"
      >
        <div
          className="catalog__movies-list"
        >
          <article
            className="small-movie-card catalog__movies-card"
          >
            <div
              className="small-movie-card__image"
              data-id={2}
              onClick={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              <img
                alt="Dardjeeling Limited"
                className="small-movie-card__pic"
                height="175"
                src="img/no-country-for-old-men.jpg"
                width="280"
              />
            </div>
            <h3
              className="small-movie-card__title"
              onClick={[Function]}
            >
              <a
                className="small-movie-card__link"
                data-id={2}
                href="movie-page.html"
              >
                Dardjeeling Limited
              </a>
            </h3>
          </article>
          <article
            className="small-movie-card catalog__movies-card"
          >
            <div
              className="small-movie-card__image"
              data-id={3}
              onClick={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              <img
                alt="Dardjeeling Limited"
                className="small-movie-card__pic"
                height="175"
                src="img/no-country-for-old-men.jpg"
                width="280"
              />
            </div>
            <h3
              className="small-movie-card__title"
              onClick={[Function]}
            >
              <a
                className="small-movie-card__link"
                data-id={3}
                href="movie-page.html"
              >
                Dardjeeling Limited
              </a>
            </h3>
          </article>
        </div>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
